File: 编译原理/TinyScript-Js/src/parser/util/PeekTokenIterator.js
const PeekIterator = require('../../common/PeekIterator')
const ParseException = require('./ParseException')

/**
 * Token流视察器
 */
class PeekTokenIterator extends PeekIterator {
  /**
   * 
   * @param {Iterator} it
   */
  constructor(it) {
    super(it)
  }

  /**
   * 检查下一个流中的符号是不是匹配一个值，并取出下一个符号
   * @param {string} value 
   * @return {Token}
   * @throws {ParseException}
   */
  nextMatchValue(value) {
    const token = this.next()
    if (token.getValue() !== value) {
      throw ParseException.fromToken(token)
    }
    return token
  }

  /**
   * 检查下一个流中的符号是不是匹配一个类型，并取出下一个符号
   * @param {string} type 
   * @return {Token}
   * @throws {ParseException}
   */
  nextMatchType(type) {
    const token = this.next()
    if (token.getType() !== type) {
      throw ParseException.fromToken(token)
    }
    return token
  }
}

module.exports = PeekTokenIterator

File: 编译原理/TinyScript/src/main/java/parser/util/PeekTokenIterator.java
package parser.util;

import common.PeekIterator;
import lexer.Token;
import lexer.TokenType;

import java.util.stream.Stream;

/**
 * Token流视察器
 */
public class PeekTokenIterator extends PeekIterator<Token> {

    public PeekTokenIterator(Stream<Token> stream) {
        super(stream);
    }

    /**
     * 检查下一个流中的符号是不是匹配一个值，并取出下一个符号
     * @param value
     * @return
     * @throws ParseException
     */
    public Token nextMatch(String value) throws ParseException {
        Token token = this.next();
        if (!token.getValue().equals(value)) {
            throw new ParseException(token);
        }
        return token;
    }

    /**
     * 检查下一个流中的符号是不是匹配一个类型，并取出下一个符号
     * @param type
     * @return
     * @throws ParseException
     */
    public Token nextMatch(TokenType type) throws ParseException {
        Token token = this.next();
        if (!token.getValue().equals(type)) {
            throw new ParseException(token);
        }
        return token;
    }
}
